src/Native/LibMFWrapper/mf.cpp

```diff
@@ -0,0 +1,1314 @@
+/*
+Copyright(c) 2014 - 2015 The LIBMF Project.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met :
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and / or other materials provided with the distribution.
+
+3. Neither name of copyright holders nor the names of its contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED.IN NO EVENT SHALL THE REGENTS OR
+CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES(INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include <algorithm>
+#include <condition_variable>
+#include <iomanip>
+#include <iostream>
+#include <numeric>
+#include <queue>
+#include <thread>
+#include <unordered_set>
+#include <random>
+#include <cstring>
+#include <fstream>
+#include <vector>
+#include <new>
+#include <string>
+#include <memory>
+
+#include "mf.h"
+
+#if defined USESSE
+#include <pmmintrin.h>
+#endif
+
+#if defined USEAVX
+#include <immintrin.h>
+#endif
+
+#if defined USEOMP
+#include <omp.h>
+#endif
+
+namespace mf
+{
+
+using namespace std;
+
+namespace
+{
+
+mf_int const kALIGNByte = 32;
+mf_int const kALIGN = kALIGNByte/sizeof(mf_float);
+
+vector<string> tokenize(string str)
+{
+    vector<string> tokens;
+
+    if(str.empty())
+        return tokens;
+
+    basic_string<char>::size_type begin = 0, end = 0;
+    while((end = str.find_first_of(" ", begin)) != string::npos)
+    {
+        string token = str.substr(begin, end-begin);
+        if(token.size() != 0)
+            tokens.push_back(token);
+        begin = end + 1;
+    }
+    string tail = str.substr(begin);
+    if(tail.size() != 0)
+        tokens.push_back(tail);
+
+    return tokens;
+}
+
+class Scheduler
+{
+public:
+    Scheduler(mf_int nr_bins, mf_int nr_threads, vector<mf_int> cv_blocks);
```

Comment:
>Scheduler [](start = 4, length = 9)

So much code, so few (none) comments.

Reply:
They are invisible now as a submodule.

---
In reply to: [225361311](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225361311) [](ancestors = 225361311)